pennylane/compiler/python_compiler/transforms/quantum/tree_traversal.py

```
@@ -0,0 +1,2477 @@
+                            # It's a new maximum depth! Clear the old list and start fresh.
+                            max_depth_ops.clear()
+                            max_depth_ops.append((child_op, new_depth))
+
+                        # 2. Is this at the same depth as the current max? (A tie)
+                        elif new_depth == max_depth_ops[0][1]:
+                            # Add it to the list of winners.
+                            max_depth_ops.append((child_op, new_depth))
+
+                    # Recursively search inside this child op (regardless of its type)
+                    # We pass the potentially *increased* new_depth.
+                    self._find_deepest_for_recursive(child_op, new_depth, max_depth_ops)
+
+    def unroll_loop(self, op: scf.ForOp, rewriter: PatternRewriter) -> None:
+        """Unroll an scf.ForOp into separate branches for each operator."""
+
+        def find_constant_bound(bound: SSAValue) -> tuple[bool, Operation | None]:
+            """Find the constant value of a SSA in ForOp bounds"""
+
+            check_bound = bound
+            while True:
+                if isinstance(check_bound.owner, arith.ConstantOp):
+                    return True, check_bound.owner
+                if isinstance(check_bound.owner, stablehlo.ConstantOp):
+                    return True, check_bound.owner
+                if isinstance(check_bound, BlockArgument):
+                    return False, None
+                if len(check_bound.owner.operands) == 0:
+                    return False, None
+
+                check_bound = check_bound.owner.operands[0]
+
+        ub_found, ub_op = find_constant_bound(op.ub)
+        lb_found, lb_op = find_constant_bound(op.lb)
+        step_found, step_op = find_constant_bound(op.step)
+
+        assert (
+            lb_found and ub_found and step_found
+        ), "UnrollLoopPattern: Cannot unroll loop, bounds or step are not constant."
+
```

Is this assertion caught by the pass infrastructure? Or will this be raised to the user? If the latter we should raise an exception explaining what went wrong, assertions are better for checking internal assumptions (which should always hold, unless there is a bug).

---

Fix it 
a612bb119